android/src/main/java/com/tailscale/ipn/IPNService.kt

```diff
@@ -27,26 +26,34 @@ open class IPNService : VpnService(), libtailscale.IPNService {
     App.getApplication()
   }
 
-  override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
-    if (intent != null && "android.net.VpnService" == intent.action) {
-      // Start VPN and connect to it due to Always-on VPN
-      val i = Intent(IPNReceiver.INTENT_CONNECT_VPN)
-      i.setPackage(packageName)
-      i.setClass(applicationContext, IPNReceiver::class.java)
-      sendBroadcast(i)
-
-    // If intent is null, the service is restarting because the system is attempting to re-create the killed service. If this is the case, check whether we are able to start the the VPN before starting.
-    } else if (intent != null || getAbleToStartVPN(this.getApplicationContext())) {
-      Libtailscale.requestVPN(this)
-      App.getApplication().setWantRunning(true)
-    }
-    return START_STICKY
-  }
-
-  override public fun close() {
+  override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int =
+      when (intent?.action) {
+        ACTION_STOP_VPN -> {
+          close()
+          START_NOT_STICKY
+        }
+        ACTION_START_VPN -> {
+          Libtailscale.requestVPN(this)
+          START_STICKY
+        }
+        "android.net.VpnService" -> {
+          // This means we were started by Android due to Always On VPN.
+          // Get the application to make sure it's been initialized, then
+          // request the VPN.
+          App.getApplication()
+          Libtailscale.requestVPN(this)
+          START_STICKY
+        }
+        else -> {
+          // This should never really happen, but we need when to be
+          // exhaustive.
+          START_NOT_STICKY
+        }
+      }
+
+  override fun close() {
     stopForeground(true)
     Libtailscale.serviceDisconnect(this)
-    App.getApplication().setWantRunning(false)
```

Comment:
Move this to `stopVPN()`. Also, if the Go side stops the VPN, it's already set wantRunning to false.